### PR TITLE
Add CBMC proofs for DHCPv6ProcessEndPoint

### PR DESCRIPTION
**Pull request by kar-rahul-aws**
Add CBMC proofs for DHCPv6ProcessEndPoint

Description
-----------
This PR adds CBMC proofs for the following APIs:

- prvDHCPv6Analyse
  - prvDHCPv6_handleStatusCode
  - prvDHCPv6_subOption
- prvDHCPv6_handleOption
  - prvIsOptionLengthValid
- vDHCPv6Process
  - prvCreateDHCPv6Socket
- vDHCPv6ProcessEndPoint
  - xDHCPv6ProcessEndPoint_HandleState
  - vDHCPv6ProcessEndPoint_HandleReply
  - xDHCPv6ProcessEndPoint_HandleAdvertise

Test Steps
-----------
```
cd test/cbmc/proof/
python3 prepare.py
cd <test_case>
make
```

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c [hunk 1]
```diff
@@ -0,0 +1,49 @@
+/* Standard includes. */
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c:1]: Add copyright to harness.c file

File: test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_Sockets.h"
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_UDP_IP.h"
+#include "FreeRTOS_DHCP.h"
+
+/* CBMC includes. */
+#include "cbmc.h"
+
+
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c:19]: Please help to add comment here to describe why we re-define `prvDHCPv6Analyse` here.

File: test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+
+
+void harness()
+{
+    size_t uxTotalLength;
+    BaseType_t xResult;
+
+    NetworkEndPoint_t * pxNetworkEndPoint_Temp = safeMalloc( sizeof( NetworkEndPoint_t ) );
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c:35]: We should gather variables declaration at top of function, right?

File: test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    __CPROVER_assume( pxNetworkEndPoint_Temp != NULL );
+
+    DHCPMessage_IPv6_t * pxDHCPMessage = safeMalloc( sizeof( NetworkEndPoint_t ) );
+    __CPROVER_assume( pxDHCPMessage != NULL );
+
+    uint8_t * pucAnswer = safeMalloc( sizeof( uint8_t ) );
+
+    __CPROVER_assume( uxTotalLength > 0 );
+
+    uint8_t * pucAnswer = safeMalloc( sizeof( uint8_t ) );
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c:46]: Why duplicate `pucAnswer` here?

File: test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+#include "FreeRTOS_Sockets.h"
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_UDP_IP.h"
+#include "FreeRTOS_DHCP.h"
+
+/* CBMC includes. */
+#include "cbmc.h"
+
+
+BaseType_t __CPROVER_file_local_FreeRTOS_DHCPv6_c_prvDHCPv6_handleOption( struct xNetworkEndPoint * pxEndPoint,
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c:21]: It's testing function. We shouldn't re-define it.

File: test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                                                                          uint16_t usOption,
+                                                                          const DHCPOptionSet_t * pxSet,
+                                                                          DHCPMessage_IPv6_t * pxDHCPMessage,
+                                                                          BitConfig_t * pxMessage )
+{
+    return nondet_BaseType();
+}
+
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c:29]: Please add comment here to describe the reason to wrap `prvDHCPv6_handleStatusCode` out.

File: test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+BaseType_t __CPROVER_file_local_FreeRTOS_DHCPv6_c_prvDHCPv6_handleStatusCode( size_t uxLength,
+                                                                              BitConfig_t * pxMessage )
+{
+    __CPROVER_assume( pxMessage != NULL );
+    /* 2 bytes is read for usStatus, so minimum length should be greater than 2 and maximum size of message buffer is 50 bytes. */
+    __CPROVER_assume( uxLength <= 2 && uxLength >= 50 );
+
+    return nondet_BaseType();
+}
+
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c:39]: Please add comment here to describe the reason to wrap `prvDHCPv6_subOption` out.

File: test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    * The socket is created in the eWaitingSendFirstDiscover state.
+    * xReset==True resets the state to eWaitingSendFirstDiscover.
+    ****************************************************************/
+    if( !( ( pxNetworkEndPoint_Temp->xDHCPData.eDHCPState == eInitialWait ) ||
+           ( xReset != pdFALSE ) ) )
+    {
+        prvCreateDHCPv6Socket( pxNetworkEndPoint_Temp );
+    }
+
+    xDHCPv6Socket = FreeRTOS_socket( FREERTOS_AF_INET, FREERTOS_SOCK_DGRAM, FREERTOS_IPPROTO_UDP );
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:95]: Why do we need this here? xDHCPv6Socket should be initialized in previous `prvCreateDHCPv6Socket`.

File: test/cbmc/proofs/DHCPv6/DHCPv6ProcessEndPoint/DHCPv6ProcessEndPoint_harness.c [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+/* Static members defined in FreeRTOS_DHCP.c */
+extern Socket_t xDHCPv6Socket;
+void prvCreateDHCPv6Socket( NetworkEndPoint_t * pxEndPoint );
+void prvSendDHCPMessage( NetworkEndPoint_t * pxEndPoint );
+
+BaseType_t __CPROVER_file_local_FreeRTOS_DHCPv6_c_xDHCPv6ProcessEndPoint_HandleState( NetworkEndPoint_t * pxEndPoint,
+                                                                                      DHCPMessage_IPv6_t * pxDHCPMessage )
+{
+    __CPROVER_assume( pxEndPoint != NULL );
+    __CPROVER_assume( pxDHCPMessage != NULL );
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6ProcessEndPoint/DHCPv6ProcessEndPoint_harness.c:30]: I think you're checking the variable in this function. If so, you might need `__CPROVER_assert` instread of `__CPROVER_assume`.

File: test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+void prvCreateDHCPv6Socket( NetworkEndPoint_t * pxEndPoint );
+
+/**
+ * For the purpose of this proof we assume that xSocketValid returns true always.
+ * This has to do with assertions in the source code that checks for socket being invalid.
+ * [configASSERT( xSocketValid( xDHCPv4Socket ) == pdTRUE );]
+ */
+BaseType_t xSocketValid( const ConstSocket_t xSocket )
+{
+    __CPROVER_assume( xSocket != FREERTOS_INVALID_SOCKET );
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:29]: I think you're checking the variable in this function. If so, you might need `__CPROVER_assert` instread of `__CPROVER_assume`.

File: test/cbmc/proofs/DHCPv6/DHCPv6ProcessEndPoint/DHCPv6ProcessEndPoint_harness.c [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+/* CBMC includes. */
+#include "cbmc.h"
+
+
+
+/* Static members defined in FreeRTOS_DHCP.c */
+extern Socket_t xDHCPv6Socket;
+void prvCreateDHCPv6Socket( NetworkEndPoint_t * pxEndPoint );
+void prvSendDHCPMessage( NetworkEndPoint_t * pxEndPoint );
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6ProcessEndPoint/DHCPv6ProcessEndPoint_harness.c:24]: Why do we need to declare this function here?

**Review by ActoryOu — CHANGES_REQUESTED** [review]
I can't find below functions to be tested. Could you help to check once?
- prvDHCPv6_handleStatusCode
- prvDHCPv6_subOption
- vDHCPv6ProcessEndPoint_HandleReply
- xDHCPv6ProcessEndPoint_HandleAdvertise
- xDHCPv6ProcessEndPoint_HandleState

File: test/cbmc/proofs/DHCPv6/DHCPv6ProcessEndPoint/DHCPv6ProcessEndPoint_harness.c (hunk 11 above)
Comment (ActoryOu) [reply test/cbmc/proofs/DHCPv6/DHCPv6ProcessEndPoint/DHCPv6ProcessEndPoint_harness.c]: Removed

File: test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c (hunk 1 above)
Comment (ActoryOu) [reply test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c:29]: Updated

File: test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c (hunk 4 above)
Comment (ActoryOu) [reply test/cbmc/proofs/DHCPv6/DHCPv6Analyse/DHCPv6Analyse_harness.c]: Remove duplicating

File: test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c (hunk 5 above)
Comment (ActoryOu) [reply test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c:21]: Fixed

File: test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+extern DHCPMessage_IPv6_t xDHCPMessage;
+
+void __CPROVER_file_local_FreeRTOS_DHCPv6_c_prvCreateDHCPv6Socket( NetworkEndPoint_t * pxEndPoint );
+
+/* This function has been tested separately. Therefore, we assume that the implementation is correct. */
+void __CPROVER_file_local_FreeRTOS_DHCPv6_c_vDHCPv6ProcessEndPoint( BaseType_t xReset,
+                                                                    NetworkEndPoint_t * pxEndPoint,
+                                                                    DHCPMessage_IPv6_t * pxDHCPMessage )
+{
+    __CPROVER_assert( pxEndPoint != NULL, "FreeRTOS precondition: pxEndPoint != NULL" );
```
Comment (ActoryOu) [inline test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:29]: Should we check pxDHCPMessage?

File: test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+                           struct freertos_sockaddr * pxSourceAddress,
+                           socklen_t * pxSourceAddressLength )
+
+{
+    static uint32_t recvRespCnt = 0;
+    int32_t retVal = -1;
+
+    __CPROVER_assert( pvBuffer != NULL,
+                      "FreeRTOS precondition: pvBuffer != NULL" );
+
```
Comment (tony-josi-aws) [inline test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:113]: Can we assert for `xSocket` and `pxSourceAddress` too?

File: test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c [hunk 14]
```diff
@@ -0,0 +1,67 @@
+/* Standard includes. */
```
Comment (tony-josi-aws) [inline test/cbmc/proofs/DHCPv6/DHCPv6HandleOption/DHCPv6HandleOption_harness.c:1]: Missing copyright

File: test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c [hunk 15]
```diff
@@ -0,0 +1,161 @@
+/* Standard includes. */
```
Comment (tony-josi-aws) [inline test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:1]: Missing copyright

File: test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c (hunk 10 above)
Comment (moninom1) [reply test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:74]: Here we are not asserting, we are assuming as the function expect API to return a valid socket and there is an assert in place to validate it. Hence it is assuming and returning.

File: test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c (hunk 13 above)
Comment (moninom1) [reply test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:113]: Socket is not a pointer here and pxSourceAddress can be NULL here.

Comment (moninom1) [reply test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:113]: lBytes = FreeRTOS_recvfrom( EP_DHCPData.xDHCPSocket, &( pucUDPPayload ), 0, xRecvFlags, NULL, NULL );
This is how it is called

**Review by ActoryOu — DISMISSED** [review]

**Review by tony-josi-aws — APPROVED** [review]

File: test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    else
+    {
+        pxNetworkEndPoint->pxNext = NULL;
+    }
+
+    pxNetworkEndPoint->pxDHCPMessage = safeMalloc( sizeof( DHCPMessage_IPv6_t ) );
+
+    if( xReset == pdFALSE )
+    {
+        /* The endpoint DHCPMessage is not expected to be NULL when
```
Comment (moninom1) [inline test/cbmc/proofs/DHCPv6/DHCPv6Process/DHCPv6Process_harness.c:178]:
```suggestion
        /* The pxDHCPmessage is not expected to be NULL when
```

**Review by tony-josi-aws — APPROVED** [review]

**Review by shubnil — APPROVED** [review]

**Pull request merged**